tests/io/test_multi_table.py

```diff
@@ -0,0 +1,241 @@
+        with pytest.warns(UserWarning, match="The table is"):
+            SpatialData(
+                shapes={
+                    "test_shapes": test_shapes["poly"],
+                },
+                tables={"shape_annotate": table2},
+            )
+        test_sdata = SpatialData(
+            shapes={
+                "test_shapes": test_shapes["poly"],
+            },
+            tables={"shape_annotate": table},
+        )
+        test_sdata.write(tmpdir)
+        sdata = SpatialData.read(tmpdir)
+
+        assert isinstance(sdata["shape_annotate"], AnnData)
+
+        assert_equal(test_sdata["shape_annotate"], sdata["shape_annotate"])
+
+        # note (to keep in the code): these tests here should silmulate the interactions from teh users; if the syntax
+        # here we are matching the table to the shapes and viceversa (= subset + reordeing)
+        # there is already a function to do one of these two join operations which is match_table_to_element()
+        # is too verbose/complex we need to adjust the internals to make it smoother
+        # # use case example 1
+        # # sorting the shapes to match the order of the table
+        # alternatively, we can have a helper function (join, and simpler ones "match_table_to_element()"
+        # "match_element_to_table()", "match_annotations_order(...)", "mathc_reference_eleemnt_order??(...)")
+        # sdata["visium0"][SpatialData.get_instance_key_column(sdata.table['visium0'])]
+        # assert ...
+        # # use case example 2
+        # # sorting the table to match the order of the shapes
+        # sdata.table.obs.set_index(keys=["__instance_id__"])
+        # sdata.table.obs[sdata["visium0"]]
+        # assert ...
+
+    def test_paired_elements_tables(self, tmp_path: str):
+        tmpdir = Path(tmp_path) / "tmp.zarr"
+        table = _get_table(region="poly")
+        table2 = _get_table(region="multipoly")
+        table3 = _get_table(region="non_existing")
+        with pytest.warns(UserWarning, match="The table is annotating elements not present in the SpatialData object"):
+            SpatialData(
+                shapes={"poly": test_shapes["poly"], "multipoly": test_shapes["multipoly"]},
+                table={"poly_annotate": table, "multipoly_annotate": table3},
+            )
+        test_sdata = SpatialData(
+            shapes={"poly": test_shapes["poly"], "multipoly": test_shapes["multipoly"]},
+            table={"poly_annotate": table, "multipoly_annotate": table2},
+        )
+        test_sdata.write(tmpdir)
+        test_sdata = SpatialData.read(tmpdir)
+        assert len(test_sdata.tables) == 2
+
+    def test_single_table_multiple_elements(self, tmp_path: str):
+        tmpdir = Path(tmp_path) / "tmp.zarr"
+        table = _get_table(region=["poly", "multipoly"])
+        subset = table[table.obs.region == "multipoly"]
+        with pytest.raises(ValueError, match="Regions in"):
+            TableModel().validate(subset)
```

Comment:
This gives the error "ValueError: Regions in the AnnData object and `region` do not match." Which is quite bad in terms of ergonomics, because we want the user to be able to subset a table without having to worry about updating the region element.

I suggest to do the following:
- we write helper functions (static methods of SpatialData or of TableModel) to access "region", "region_key", "instance_key"; in the tutorial we will show that this is what the user should call.
- we never populate adata.uns['spatialdata_attrs']['region'], instead we compute it on the fly when the respective of such methods is called, which simply gives the levels of the categorical column specified by "region_key". If such column is not categorical we compute the unique values and we tell the user that that column should be a categorical.
- when we write, we compute `region` and write it in the Zarr metadata, but we never read it in memory (it's for other tools or for convenient remote access).

Reply:
Hmm I think this is something to get feedback on perhaps in a community meeting.

Reply:
> This gives the error "ValueError: Regions in the AnnData object and region do not match." Which is quite bad in terms of ergonomics, because we want the user to be able to subset a table without having to worry about updating the region element.

I think I disagree, as in it's true it is not ergonomic, but also this table was modified outside of spatialdata and hence it must be validated again before assigning to spatialdata. I think that this type of ergonomics could very well be pushed upstream in anndata. I would keep the error as of now and consider to update the ergonomics in spatialdata

Reply:
> hence it must be validated again before assigning to spatialdata. 
Here the problem is triggered by the validation, I would not expect the new object to be invalid just by subsetting. 

I don't think something upstream in AnnData would help here because `region` it's something spatialdata specific.

I think the approach I described is quite ergonomic, we can discuss it maybe at the community meeting today.

Reply:
how do we know which is "region_key" a priori? we don't know that.

Reply:
Anyway, @giovp and I discussed this in the community meeting, and @melonora and I just discussed this in a call. We agreed on addressing the removal of the region logic from memory (will keeping it on disk) in a follow up PR. Issue to track this here: https://github.com/scverse/spatialdata/issues/419. For the moment using a warning instead of an error is a gentler way to handle this.